src/window_manipulation.rs

```diff
@@ -0,0 +1,374 @@
+				let dimensions = (
+					// x
+					if diff_x > 0 {
+						// If `diff_x` is positive, we can just add the numbers.
+						orig_size.0 as u32 + diff_x as u32
+					} else {
+						// If `diff_x` is negative, we need to make sure it won't shrink the width
+						// too far.
+						if (-diff_x as u16) < orig_size.0 {
+							// If `diff_x` decreases the width by less than the width itself, we
+							// can subtract it.
+							orig_size.0 as u32 - -diff_x as u32
+						} else {
+							// If `diff_x` shrinks the width too much, we'll set the width to 1.
+							1
+						}
+					},
+					// y
+					if diff_y > 0 {
+						// If `diff_y` is positive, we can just add the numbers.
+						orig_size.1 as u32 + diff_y as u32
+					} else {
+						// If `diff_y` is negative, we need to make sure it won't shrink the
+						// height too far.
+						if (-diff_y as u16) < orig_size.1 {
+							// If `diff_y` decreases the height by less than the height itself, we
+							// can substract it.
+							orig_size.1 as u32 - -diff_y as u32
+						} else {
+							// If `diff_y` shrinks the height too much, we'll set the height to 1.
+							1
+						}
+					},
+				);
+
+				// TODO: This sends far too many requests! We need to make sure that not too many
+				//       changes to the window's dimensions are sent, or the X server melts.
+				set_dimensions(conn, window, dimensions);
+			}
+		}
+
+		Ok(())
+	}
+}
+
+/// Sends a [GetGeometry](xcb::x::GetGeometry) request for the given window and returns its reply.
+///
+/// Does not flush the connection.
+///
+/// TODO: Move out to utility module.
+fn get_geometry(conn: &Connection, window: Window) -> x::GetGeometryCookie {
+	trace!(window = window.resource_id(), "Requesting window geometry");
+	conn.send_request(&x::GetGeometry {
+		drawable: x::Drawable::Window(window),
+	})
+}
+
+/// Sends a [ConfigureWindow](xcb::x::ConfigureWindow) request to change the coordinates of the
+/// given window.
+///
+/// Does not flush the connection.
+///
+/// TODO: Move out to utility module.
+fn set_position(conn: &Connection, window: Window, coords: (i32, i32)) {
+	trace!(
+		window = window.resource_id(),
+		x = coords.0,
+		y = coords.1,
+		"Configuring window coordinates"
+	);
+	conn.send_request(&x::ConfigureWindow {
+		window,
+		value_list: &[ConfigWindow::X(coords.0), ConfigWindow::Y(coords.1)],
+	});
+}
+
+/// Sends a [ConfigureWindow](xcb::x::ConfigureWindow) request to change the dimensions of the
+/// given window.
+///
+/// Does not flush the connection.
+///
+/// TODO: Move out to utility module.
+fn set_dimensions(conn: &Connection, window: Window, dimensions: (u32, u32)) {
+	trace!(
+		window = window.resource_id(),
+		width = dimensions.0,
+		y = dimensions.1,
+		"Configuring window dimensions"
+	);
+	conn.send_request(&x::ConfigureWindow {
+		window,
+		value_list: &[
+			ConfigWindow::Width(dimensions.0),
+			ConfigWindow::Height(dimensions.1),
+		],
+	});
+}
```

Comment:
As the TODOs say here, these should definitely be moved out into a utility module, along with `crate::init_window`.

Reply:
Done.